etcd-client/src/watch/mod.rs

```diff
@@ -40,163 +40,545 @@
 //!
 //! ```
 
-use std::sync::Arc;
+use std::collections::{BTreeMap, HashMap};
+use std::sync::{Arc, Weak};
+use std::time::Duration;
 
-use async_trait::async_trait;
+use async_broadcast::{InactiveReceiver, Receiver as BroadcastRx, Sender as BroadcastTx};
+use async_std::channel::bounded;
+use crossbeam_queue::SegQueue;
 use futures::future::FutureExt;
-
 use futures::stream::StreamExt;
-use smol::channel::{unbounded, Receiver, Sender};
-use smol::stream::Stream;
+use futures::{SinkExt, TryFutureExt};
+use grpcio::{StreamingCallSink, WriteFlags};
+use smol::channel::{unbounded, Sender};
+use smol::Task;
 
 pub use watch_impl::{EtcdWatchRequest, EtcdWatchResponse};
 
-use futures::prelude::*;
-
-use crate::lazy::{Lazy, Shutdown};
+use crate::lazy::Lazy;
 use crate::protos::kv;
-use crate::protos::rpc;
+use crate::protos::rpc::WatchRequest;
 use crate::protos::rpc_grpc::WatchClient;
 use crate::EtcdKeyValue;
 use crate::KeyRange;
 use crate::Result;
-use grpcio::WriteFlags;
 
 /// Watch implementation mod.
 mod watch_impl;
 
+/// the timeout of waiting etcd response
+const WATCH_REQUEST_TIMEOUT_SEC: u64 = 2;
+
+/// watch id
+type WatchID = i64;
+
+/// the recorded watch info in a lock
+#[derive(Default)]
+struct WatchedMap {
+    /// keyrange to watchid
+    keyrange_2_watchid: BTreeMap<KeyRange, WatchID>,
+    /// watchid to (sender to user, user receiver)
+    watchid_2_detail: HashMap<
+        WatchID,
+        (
+            BroadcastTx<Option<EtcdWatchResponse>>,
+            Weak<SingleWatchEventReceiverInner>,
+        ),
+    >,
+}
+impl WatchedMap {
+    /// add new watched info
+    fn add_watched(
+        &mut self,
+        watchid: WatchID,
+        keyrange: KeyRange,
+        sender_2_user: BroadcastTx<Option<EtcdWatchResponse>>,
+        user_receiver: Weak<SingleWatchEventReceiverInner>,
+    ) {
+        self.keyrange_2_watchid.insert(keyrange, watchid);
+        self.watchid_2_detail
+            .insert(watchid, (sender_2_user, user_receiver));
+    }
+    /// remove the watch to cancel
+    fn remove_watch(&mut self, keyrange: &KeyRange) -> Option<WatchID> {
+        if let Some(watch_id) = self.keyrange_2_watchid.remove(keyrange) {
+            log::debug!("removed watch");
+            drop(
+                self.watchid_2_detail
+                    .remove(&watch_id)
+                    .unwrap_or_else(|| panic!("")),
+            );
+            return Some(watch_id);
+        }
+        None
+    }
+
+    /// get sender 2 user to send watched event
+    fn get_sender_2_user(
```

Comment:
The name of this function and the comments are confusing. Can we make it more clear?

Reply:
How about `get_broadcast_sender_2_watcher`